modules/tinymce/src/plugins/quickbars/main/ts/insert/Toolbars.ts

```diff
@@ -13,9 +13,15 @@ const addToEditor = (editor: Editor): void => {
         const sugarNode = SugarElement.fromDom(node);
         const textBlockElementsMap = editor.schema.getTextBlockElements();
         const isRoot = (elem: SugarElement<Node>) => elem.dom === editor.getBody();
+        const isFakeCaret = (elem: SugarElement<Node>) =>
+          Attribute.has(elem, 'data-mce-bogus') || SelectorExists.ancestor(elem, '[data-mce-bogus="all"]', isRoot);
+
         return SelectorFind.closest(sugarNode, 'table', isRoot).fold(
-          () => PredicateFind.closest(sugarNode, (elem) =>
-            SugarNode.name(elem) in textBlockElementsMap && editor.dom.isEmpty(elem.dom), isRoot).isSome(),
+          () => PredicateExists.closest(
+            sugarNode,
+            (elem) => !isFakeCaret(elem) && SugarNode.name(elem) in textBlockElementsMap && editor.dom.isEmpty(elem.dom),
```

Comment:
I think the order of these calculations could be improved. Let's break it down: 

* isFakeCaret is potentially going to look all the way up the tree to the root.
* SugarNode is an O(1) operation

So I think the first improvement would be to do the SugarNode.name check first, as that will potentially prevent you from needing to do any "up the dom searching" with isFakeCaret. Remember, that for the common case of the current element **doesn't have** `data-mce-bogus`, `isFakeCaret` is going to search all the way up to the root on *every* PredicateExists call. That's a lot of repeated effort, and I don't think the result is ever going to change if it's false.

Then, you could probably combine the two searches so that we aren't looking up every step. Or you could look at doing a Thunk.cached for the SelectorExists.ancestor. It just seems that there is something inefficient here, but I might be misreading the code.

Reply:
I moved `isFakeCaret` so it is the last check

Reply:
@tiny-ben-tran this only does the first part of what Morgan mentioned. Was there a reason the second part was skipped?

Reply:
Thanks Lee, I totally missed it.

I combined the two searches as suggested.

Reply:
One thing to remember here is that the predicate function gets called multiple times for every node up the tree from the selection.

Reply:
Yeah, my read of the original code is what we *want* to happen is this: 

* if anything up the tree is a table, then do not render
* otherwise, if we find an ancestor text block-capable node up the tree, and it's empty, render. Otherwise, keep looking.

The additional things to add are: 

* if anything up the tree is a data-mce-bogus=all, then do not render
* otherwise, if we find an ancestor that has data-mce-bogus, then do not render.

So it seems like we have two different types of checks. My instinct suggests you should be able to consolidate them. So the first `SelectorFind.closest` could look for either table, or [data-mce-bogus=all], as both conditions will stop it from rendering. Otherwise, you run through the PredicateExists.closest and then you'll only need to check for the data-mce-bus attribute and the other things it was already checking for.

Is that preserving the right intent @lnewson ?

Reply:
Sorry I missed this, but yes that sounds correct to me.

Reply:
Should the toolbar be shown if the cursor is placed in the deepest span element of `<p><span><span data-mce-bogus="1"><span>{cursor here}</span></span></span></p>`?

Reply:
@lnewson @ephox-mogran, can you give this PR another review?

Reply:
LGTM 👍